File: semestr-8/Jezyk C#/alexander/Csharp/Archiwum/2021s/Id06TC2/Decorator/Decorator/CarDecorator.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Decorator
{
    public abstract class CarDecorator:Car
    {
        private Car car;
        public CarDecorator(Car car) { this.car = car; }
        public virtual int price() { return car.price(); }
        public virtual string info() { return car.info(); }
        public Car baseCar { get { return car is CarDecorator?((CarDecorator)car).baseCar:car; } }
    }
    public abstract class CarDecoratorFactory
    {
        private delegate Car maker(Car car);
        private static Dictionary<string, maker> map = new Dictionary<string, maker>()
        {
            {"Air Condition",(car)=>new AirCondition(car)},
            {"Leather Seats",(car)=>new LeatherSeats(car)},
            {"Tinted Window",(car)=>new TintedWindow(car)},
            {"Winter Tire",(car)=>new WinterTire(car)},
        };
        public static List<string> Kinds { get { return map.Keys.OrderBy((kind) => kind).ToList(); } }
        public static Car Make(string kind, Car car) { return map[kind](car); }
    }
    public class AirCondition:CarDecorator
    {
        public AirCondition(Car car) : base(car) { }
        public override int price() 
        {
            return base.price() + (baseCar is Mercedes ? 5300 : 5000);
        }
        public override string info() { return base.info()+" + Air Condition"; }
    }
    public class LeatherSeats : CarDecorator
    {
        public LeatherSeats(Car car) : base(car) { }
        public override int price() { return base.price() + 9000; }
        public override string info() { return base.info() + " + Leather Seats"; }
    }
    public class TintedWindow : CarDecorator
    {
        public TintedWindow(Car car) : base(car) { }
        public override int price() { return base.price() + baseCar.tintedWindowPrice(); }
        public override string info() { return base.info() + " + Tinted Window"; }
    }
    public class WinterTire : CarDecorator
    {
        private static Dictionary<string, int> pricelist = new Dictionary<string, int>()
        {
            {typeof(Mercedes).Name,3900},
            {typeof(Volvo).Name,4500},
        };
        public WinterTire(Car car) : base(car) { }
        public override int price() 
        {
            int add=4000;
            try { add = pricelist[this.GetType().Name]; } catch(Exception) {}
            return base.price() + add; 
        }
        public override string info() { return base.info() + " + Winter Tire"; }
    }
}

File: semestr-8/Programowanie obiektowe w Java/alexander/POJ/Archiwum/2021s/IZ08IO1/Decorator/src/CarDecorator.java

import java.util.HashMap;
import java.util.List;
import java.util.stream.Collectors;

public abstract class CarDecorator extends Car
{
   private static HashMap<String,CarDecoratorMaker> map=new HashMap<String,CarDecoratorMaker>()
   {
      {
         put("Air Condition",(car)->new AirCondition(car));
         put("Tinted Window",(car)->new TintedWindow(car));
         put("Winter Tire",(car)->new WinterTire(car));
         put("Leather Seats",(car)->new LeatherSeats(car));
      }
   };
   public static final List<String> keys=map.keySet().stream().sorted().collect(Collectors.toList());
   public static Car make(String kind,Car car) { return map.get(kind).make(car); }
   private Car car;
   public CarDecorator(Car car) { this.car=car; }
   @Override public String info() { return car.info(); }
   @Override public int price() { return car.price(); }
}

interface CarDecoratorMaker
{
   Car make(Car car);
}